src/async_read.rs

```diff
@@ -0,0 +1,80 @@
+use bytes::BufMut;
+use futures01::{Async, Future, IntoFuture, Poll};
+use std::io::Read;
+use tokio01::io::AsyncRead;
+
+/// A AsyncRead combinator which reads from a reader until a future resolves.
+///
+/// This structure is produced by the [`AsyncReadExt::read_until`] method.
+#[derive(Clone, Debug)]
+pub struct ReadUntil<S, F, O> {
+    reader: S,
+    until: F,
+    until_res: Option<O>,
+    free: bool,
+}
+
+/// This `AsyncRead` extension trait provides a `read_until` method that terminates the reader once
+/// the given future resolves.
+pub trait AsyncReadExt: AsyncRead {
+    /// Read data from this reader until the given future resolves.
+    ///
+    /// If the future produces an error, the read will be allowed to continue indefinitely.
+    fn read_until<U, O>(self, until: U) -> ReadUntil<Self, U::Future, O>
+    where
+        U: IntoFuture<Item = O, Error = ()>,
+        Self: Sized,
+    {
+        ReadUntil {
+            reader: self,
+            until: until.into_future(),
+            until_res: None,
+            free: false,
+        }
+    }
+}
+
+impl<S> AsyncReadExt for S where S: AsyncRead {}
+
+impl<S, F, O> Read for ReadUntil<S, F, O>
+where
+    S: Read,
+{
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        self.reader.read(buf)
+    }
+}
+
+impl<S, F, O> AsyncRead for ReadUntil<S, F, O>
+where
+    S: AsyncRead,
+    F: Future<Item = O, Error = ()>,
```

Comment:
Since it doesn't seem like we really want an error here, can we maybe use [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) for `Error`?

Reply:
We can't since`ShutdownSignal` isn't infallible.